doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+As described in this document, `Dataset` should be able to replace all of our workspace types (for `MDEventWorkspace` there is not significant advantage to using the new design since data in `MDEventWorkspace` has a tree structure).
+This includes not just `MatrixWorkspace` and its child classes `Workspace2D` and `EventWorkspace` but also `TableWorkspace`, `PeaksWorkspace`, and `MDHistoWorkspace`.
+In addition, `Dataset` can likely also replace `Histogram`, `EventList`, and data structure introduced as part of **Instrument-2.0**.
+`Dataset` provides a *uniform interface* in a *single type* for all of these.
+Furthermore, `Dataset` covers many other cases that are currently impossible to represent in a single workspace or in an intuitive manner.
+
+### <a name="overview-examples"></a>Examples
+
+We begin with a series of examples.
+If clarification is needed, please first read the [following section](#overview-elements) for a description of the nomenclature and underlying objects.
+Labels in the figures use a Python-like syntax and naming is preliminary.
+
+##### A basic dataset containing two variables, a coordinate and data
+
+<img src="dataset-2d.png" height="48%" width="48%">
+
+##### Coordinates can be bin-edges, i.e., their extent can be longer by one than the corresponding dimensions of other variables
+
+<img src="dataset-2d-bin-edges.png" height="48%" width="48%">
+
+##### Coordinates do not need to be one-dimensional, i.e., they can represent a non-shared "X", as we are used to from `API::MatrixWorkspace`
+
+<img src="dataset-2d-coord-not-shared.png" height="48%" width="48%">
+
+##### More coordinates can be added, e.g., for other dimensions
+
+<img src="dataset-2d-more-coords.png" height="70%" width="70%">
+
+##### Multiple coordinates for the same dimension are also possible, as long as the coordinate tag is different, e.g., the final energy `Ef` in a spectrometer
+
+<img src="dataset-2d-multiple-coordinates-same-dimension.png" height="70%" width="70%">
+
+##### Variables can be zero-dimensional
+
+<img src="dataset-2d-variable-0d.png" height="70%" width="70%">
+
+##### Variables can have more dimensions if needed, e.g., for multi-period data or scan data
+
+<img src="dataset-3d.png" height="70%" width="70%">
+
+##### Multiple data variables of the same type are supported by a different name
+
+<img src="dataset-3d-two-data-variables.png" height="70%" width="70%">
+
+
+### <a name="overview-elements"></a>Elements
+
+`class Dataset` is the top level entity.
+It simply set of variables, instances of `class Variable`.
+`Variable` contains:
+- Tag and name identifying the variable.
+- Unit of the data held by the variable.
+- Dimensions of the data held by the variable, i.e., the side lengths of a multi-dimensional cube of data, using `class Dimensions`.
+- Data, stored in an array-like container.
+
+The dimensions of the variables in a dataset can differ, but the extents of the individual dimensions must match.
```

Comment:
We *will* support what is called "ragged"-workspace in Mantid right now. That is, there can be a *different* X-axis for each spectrum. Saying it in `Dataset` terms: Coordinates can have more than one dimensions (this may actually be another limitation of `xarray` that I should list).

However, we do *not* support a different axis *length* (this is something that was added to Mantid only less than a year ago, I believe).